### PR TITLE
remove series calling private method, fixing outer merge

### PR DESCRIPTION
**Pull request by KathiaBarahona**
Refactoring of `_get_column_name_expr_dtype` and `_determine_result_columns`.  This solves a problem we have when doing outer joins, since we are considering only left table values when having conflicting on columns. This way we also remove the hack of calling `df._index_merge` in Series class.

Let me know what do you think about always applying `COALESCE` on columns involved in the merge `on`.  Just want to avoid passing the how on multiple functions just to evaluate if we are doing `how == How.outer`. Not sure if doing this might be actually inefficient. 

Still missing changing the code from `DataFrame._index_merge`

### CODE REVIEW

File: bach/bach/merge.py [hunk 1]
```diff
@@ -126,33 +127,36 @@ def _determine_result_columns(
...
-    new_data_list += _get_column_name_expr_dtype(right_data, conflicting, suffixes[1], 'r')
+
+    left_df = left.copy()
+    right_df = right.copy() if isinstance(right, DataFrame) else right.to_frame()
+
+    conflicting_on = {l_on for l_on, r_on in zip(left_on, right_on) if l_on == r_on}
+    conflicting = (
+        (set(left_df.index) | set(left_df.data)) & (set(right_df.index) | set(right_df.data))
+    )
+    # don't add suffixes to on columns
```
Comment (KathiaBarahona) [inline bach/bach/merge.py:140]:
```suggestion
    # don't add suffixes to conflicted on columns
```

File: bach/bach/dataframe.py [hunk 2]
```diff
@@ -992,43 +992,38 @@ def _index_merge(
...
-                        right_index=True,
-                        how=how,
-                        suffixes=('', '__remove'))
+        # align index names, this way we have all matched indexes in a single series after merge
+        aligned_index = {index_name: value.index[value_index_name].copy_override(name=index_name)}
+        other = value.copy_override(
+            index=aligned_index,
+            name=key,
+            group_by=GroupBy(list(aligned_index.values())) if value.group_by else None,
+        )
```
Comment (KathiaBarahona) [inline bach/bach/dataframe.py:1013]: In a future this can be replaced by `Series.rename`, although renaming of indexes is still not supported

File: bach/bach/series/series.py [hunk 3]
```diff
@@ -504,19 +504,53 @@ def _get_supported(
...
+        ):
+            raise ValueError('dtypes of indexes to be merged should be the same')
+
+        # align index names, this way we have all matched indexes in a single series
+        other_cp = other.copy_override(
+            index={
+                caller_idx.name: other_idx.copy_override(name=caller_idx.name)
+                for caller_idx, other_idx in zip(self.index.values(), other.index.values())
+            }
+        )
```
Comment (KathiaBarahona) [inline bach/bach/series/series.py:538]: This block actually enables merging with multi-level, please let me know if it's ok, else I'll just change the condition to support only in 1 level

File: bach/bach/merge.py [hunk 4]
```diff
@@ -166,54 +170,49 @@ def _check_no_column_name_conflicts(result_columns: List[ResultSeries]):
...
-        left: DataFrame,
-        right: DataFrameOrSeries,
-        how: str,
-        on: Union[str, List[str], None],
-        left_on: Union[str, List[str],  None],  # todo: also support array-like arguments?
-        right_on: Union[str, List[str], None],
-        left_index: bool,
-        right_index: bool,
-        suffixes: Tuple[str, str]
+    left: DataFrame,
```
Comment (thijs-obj) [inline bach/bach/merge.py:207]: I'd prefer it if we do cosmetic changes to do all functions in a file or, only the ones that are actually modified. Changing indenting on some functions makes the PR bigger, without bringing consistency to the file

File: bach/bach/series/series.py [hunk 5]
```diff
@@ -504,19 +504,53 @@ def _get_supported(
...
-                    return cast('Series', df[self.name]), cast('Series', df['__other'])
-                    # todo pandas: if name of both series are same, use that name of result. we always use
-                    #  name of self
+                    return self._index_merge(other)
 
         if other.dtype.lower() not in supported_dtypes:
             raise TypeError(f'{operation_name} not supported between {self.dtype} and {other.dtype}.')
         return self, other
 
+    def _index_merge(self, other: 'Series') -> Tuple['Series', 'Series']:
```
Comment (thijs-obj) [inline bach/bach/series/series.py:513]: bit long example, but this works on main and not on this branch.
The gist of it is that if the `groupb_y` on the two `Series` don't match, then it gives an error.

```python
# Setup
from bach import DataFrame
from tests.functional.bach.test_data_and_utils import DB_TEST_URL
import sqlalchemy
import pandas as pd
import numpy as np

engine = sqlalchemy.create_engine(DB_TEST_URL)
pdf1 = pd.DataFrame(data={'a': [1, 2, 3], 'b': [2, 2, 2], 'c': [3, 3, 3]},)
pdf1 = pdf1.set_index('a')
pdf2 = pd.DataFrame(data={'x': [1, 2, 3, 4], 'y': [2, 2, 4, 4], 'z': [1, 2, 3, 4]},)
pdf2 = pdf2.set_index('x')
df1 = DataFrame.from_pandas(engine=engine, df=pdf1.reset_index(drop=False), convert_objects=True, materialization='table', if_exists='replace', name='test_table1')
df2 = DataFrame.from_pandas(engine=engine, df=pdf2.reset_index(drop=False), convert_objects=True, materialization='table', if_exists='replace', name='test_table2')
df1 = df1.reset_index(drop=True)
df1 = df1.set_index('a')
df2 = df2.reset_index(drop=True)
df2 = df2.set_index('x')

# works on main, not on current branch
xx = df1.c + df2.groupby('y')['z'].sum()
```

File: bach/bach/merge.py [hunk 6]
```diff
@@ -166,54 +170,49 @@ def _check_no_column_name_conflicts(result_columns: List[ResultSeries]):
...
+        table_alias = 'l' if suffix == suffixes[0] else 'r'
+        for series_name, series in source_series.items():
+            new_name = series_name
+            expr = series.expression.resolve_column_references(table_alias)
+
+            if series_name in conflicting_on:
+                if table_alias == 'r':
+                    continue
+                r_expr = right_series[series_name].expression.resolve_column_references('r')
+                expr = Expression.construct(f'COALESCE({expr.to_sql()}, {r_expr.to_sql()})')
```
Comment (thijs-obj) [inline bach/bach/merge.py:192]: I think just always doing coalesce here is OK.
We can later optimize in different paths for 'left', 'right' 'inner', 'cross', and 'outer' if we want to.

**Review by thijs-obj — CHANGES_REQUESTED** [review]
Generally the changes in `merge.py` are a nice improvement I think.

**mssing: tests**
I do think we should have some tests for the cases that we previously didn't support and that now work. Otherwise we'll just get regressions in the future.

**other**
Also left some (mostly small) comments on other things

File: bach/bach/series/series.py (hunk 3 above)
Comment (thijs-obj) [reply bach/bach/series/series.py:538]: Seems ok to me. @hendrik-obj  ?

Comment (hendrik-obj) [reply bach/bach/series/series.py:538]: yes, looks like it's working. good feature.

File: bach/tests/functional/bach/test_df_merge.py [hunk 7]
```diff
@@ -3,11 +3,12 @@
 """
 from decimal import Decimal
 
+import pandas as pd
 import pytest
 
 from bach import DataFrame
 from tests.functional.bach.test_data_and_utils import get_bt_with_test_data, get_bt_with_food_data, \
-    assert_equals_data, get_bt_with_railway_data
+    assert_equals_data, get_bt_with_railway_data, get_from_df
```
Comment (hendrik-obj) [inline bach/tests/functional/bach/test_df_merge.py:11]: There is an, admittedly untested, situation where the behaviour is now different from main.
When merging a series to a data frame using the name of the index of the series. ie:
```
bt = get_bt_with_test_data(True)
bt.reset_index().merge(bt.city, on='_index_skating_order')
```
Also goes wrong when series have different base nodes. Maybe a bit of an edge case but we should either not support it with an exception or implement it (and test it).

**Review by hendrik-obj — COMMENTED** [review]
Nice improvement overall.
One comment about an (untested) scenario that gives an error.

File: bach/bach/series/series.py (hunk 5 above)
Comment (KathiaBarahona) [reply bach/bach/series/series.py:513]: ouch! thanks, seems I gotta rename the indexes in the GroupBy

File: bach/tests/functional/bach/test_df_merge.py (hunk 7 above)
Comment (KathiaBarahona) [reply bach/tests/functional/bach/test_df_merge.py:11]: uh hah! actually its failing because `_index_skating_order` is in the index for `bt.city` series.

File: bach/bach/dataframe.py [hunk 8]
```diff
@@ -981,54 +981,48 @@ def __setitem__(self,
...
         Internal method used by __setitem__ to set a series using a merge on index. Modifies the DataFrame
         with the added column. The DataFrames index name is the same as the original DataFrame's.
 
         :param key: name of the column to set.
         :param value: Series that is set.
-        :param how: 'left' or 'outer'.
         """
-
-        # todo This method's functionality will be implementd in merge()
+        # todo This method's functionality will be implemented in merge()
```
[the remark this answers is not in the document]
Comment (KathiaBarahona) [reply bach/bach/dataframe.py:992]: I agree, since it's only used when setting items. Can't find other place we use it. I'll remove the comment

**Review by thijs-obj — COMMENTED** [review]
As discussed yesterday: would be great to have a specific test for what failed first and works now, the scenario of `merge(how=outer/right)` with null values in the left dataframe.

**Review by thijs-obj — APPROVED** [review]
